### PR TITLE
Add MOI support 

### PR DESCRIPTION
**Pull request by frapac**
* Add support to the new [KNITRO interface](https://www.artelys.com/tools/knitro_doc/2_userGuide.html)
* Add support to MathOptInterface
* Add comprehensive examples and unit-tests 

Currently, we are able to run a comprehensive subset of MOI Tests: 
https://github.com/Artelys/KNITRO.jl/blob/fp/moi/test/MOI_wrapper.jl
but coverage could be better (for instance by supporting MOI's SOCP problems).  

Remaining issues  
-----------------
* [ ] Add support of KNITRO's SOCP through MOI (currently, the Test `soc1test` retrieves a weird segfault...) ;
* [ ] Add support of different KNITRO's callbacks inside MOI (see #92) ;
* [ ] DRY the code of `src/MOI_wrapper.jl` with macros to avoid duplicancy ;
* [ ] Add support of change in objective after a solve (see for instance [this test](https://github.com/JuliaOpt/MathOptInterface.jl/blob/master/src/Test/contlinear.jl#L89)). Currently, KNITRO.jl returns an error. A solution would be to copy the model inside a new model and apply the change directly onto this new model.

### CODE REVIEW

**Comment by blegat** [comment]
> A solution would be to copy the model inside a new model and apply the change directly onto this new model.

If KNITRO does not support changing the objective, it is better not to support it in the wrapper. The CachingOptimizer will only apply the change in the cache and stop maintaining the KNITRO copy. Then at `MOI.optimize!`, it will copy the cache into the KNITRO model just before calling `MOI.optimize!` on it.

**Comment by frapac** [comment]
I still have a partial comprehension of the different operations allowed by `CachingOptimizer`. I should have a deeper look inside the MOI code!

**Comment by frapac** [comment]
> 
> If KNITRO does not support changing the objective, it is better not to support it in the wrapper. The CachingOptimizer will only apply the change in the cache and stop maintaining the KNITRO copy. Then at `MOI.optimize!`, it will copy the cache into the KNITRO model just before calling `MOI.optimize!` on it.

This issue is now solved by this [commit](https://github.com/JuliaOpt/KNITRO.jl/pull/94/commits/f01d1d7c77f4b14dd16da3209d3e1beefa9aff98).

File: src/KNITRO.jl [hunk 1]
```diff
@@ -16,19 +15,6 @@ module KNITRO
     @static if Sys.islinux() const libknitro = "libknitro" end
```
[the remark this answers is not in the document]
Comment (frapac) [reply src/KNITRO.jl]: Thanks for the feedback! This issue is solved by #96

**Pull request closed without merging**